### PR TITLE
BBS-128 search quoting

### PR DESCRIPTION
**Pull request by danquah**
This is mainly laying the ground-work for moving the cql-specific quoting into the opensearch provider.

### CODE REVIEW

File: modules/ting/src/Search/TingSearchRequest.php [hunk 1]
```diff
@@ -209,6 +216,35 @@ public function getSimpleQuery() {
...
+   * @param string $free_text_query
+   *   Any string.
+   *
+   * @return TingSearchRequest
+   *   the current query object.
+   */
+  public function setFreeTextQuery($free_text_query) {
+    if (!empty($free_text_query)) {
+      $this->freeTextQuery = $free_text_query;
+    }
```
Comment (kasperg) [inline modules/ting/src/Search/TingSearchRequest.php]: Is there any reason behind this check? I cannot spot is in any of the other setters.

**Review by kasperg — CHANGES_REQUESTED** [review]
This introduces another query type bringing us up to three: Simple, raw and now free.

I would like to see a better explanation of the differences and/or when to use each type than currently presented.

File: modules/opensearch/includes/opensearch.search.inc [hunk 2]
```diff
@@ -142,6 +147,10 @@ function opensearch_search_search(TingSearchRequest $ting_query) {
     $options['facets'] = $ting_query->getFacets();
   }
 
+  if ($ting_query->getTermsPerFacet() !== NULL) {
+    $options['numFacets'] = $ting_query->getTermsPerFacet();
+  }
```
Comment (kasperg) [inline modules/opensearch/includes/opensearch.search.inc]: This change does not seem to be related to quoting? (Although I do not disageee.

Comment (danquah) [reply modules/opensearch/includes/opensearch.search.inc]: Ah, one slipped trough, will fix

File: modules/opensearch/includes/opensearch.search.inc [hunk 3]
```diff
@@ -99,16 +99,13 @@ function opensearch_search_search(TingSearchRequest $ting_query) {
...
-      $query_parts[] = $simple_query_string;
-    }
-    else {
-      // Add wildcard to the search string.
-      $query_parts[]  = _ting_search_quote($simple_query_string) . '*';
+  // Add a general quoted free text search.
+  if (!empty($free_text_query = $ting_query->getFullTextQuery())) {
+    $free_text_query = _ting_search_quote($ting_query->getFullTextQuery());
+    if ($ting_query->isFuzzy()) {
+      $free_text_query = '*' . $free_text_query . '*';
```
Comment (kasperg) [inline modules/opensearch/includes/opensearch.search.inc:106]: This adds a wildcard before and after the query. The original implementation only had a wildcard after. Is this change intentional?

File: modules/opensearch/includes/opensearch.search.inc [hunk 4]
```diff
@@ -385,3 +382,38 @@ function opensearch_search_map_common_fields() {
...
+ * to look up by id.
+ *
+ * The provider can also choose to support advanced provider-specific querying.
+ *
+ * @param string $query_string
+ *   The user-provided search query.
+ *
+ * @return \Ting\Search\TingSearchRequest
+ *   The prepared query object.
+ */
```
Comment (kasperg) [inline modules/opensearch/includes/opensearch.search.inc:402]: I like the level of commenting here. How about moving it to a reference file e.g. `search.api.php` within the `ting` module file for reference and change implementations to: `Implements hook_search_prepare_reference_query`

On the other hand I know that it is not exactly a Drupal hook implementation but a Ctools plugin so...

**Review by kasperg — APPROVED** [review]
:+1: I like it. I have two questions but they may not be worth addressing here.

File: modules/opensearch/includes/opensearch.search.inc (hunk 3 above)
Comment (danquah) [reply modules/opensearch/includes/opensearch.search.inc:106]: Yes, intentional - but I should probably test how it actually affects the UX of the reference widget.

My thinking was that it makes more sense surround with the wildcard for a setting named "isFuzzy"

File: modules/opensearch/includes/opensearch.search.inc (hunk 4 above)
Comment (danquah) [reply modules/opensearch/includes/opensearch.search.inc:402]: My understanding was that that is basically the role of connie? That is, it should be an example of how a provider should handle its inputs and outputs.
